Plans: Remove `siteSpecificPlansDetailsList` and `siteSpecificPlansDetailsMixin` in favor of new actions/reducers

drewblaisdell opened the pull request:
This PR updates the components in `client/my-sites/plans/` to use site plans data in the global state.

Previously, the `/plans` components used `siteSpecificPlansDetailsList`, which:

- Was a module containing a singleton class that fetched/stored data related to site plans.
- Used `site.domain` instead of `site.slug` to namespace plans, which causes namespace conflicts for sites that share a `domain` property.
- Was used with a mixin, `siteSpecificPlansDetailsMixin`.

This PR:

- Removes `siteSpecificPlansDetailsList` and `siteSpecificPlansDetailsMixin`.
- Updates all instances of use of the former to use the new global store.

**Testing**
The plans components are used in both `/plans` and on the plans step in signup, so it is important that we assert that the behavior of both places remains unchanged.

*Signup*
- Visit http://calypso.localhost:3000/start
  - Assert that you can complete signup with the free plan, and are not redirected to checkout afterward.
  - Assert that you can complete signup with a paid plan, and are redirected to checkout afterward.
  - Assert that you can view the 'Compare Plans' page on the plans step by clicking one of the 'Learn more' links.

*Plans*
- Visit http://calypso.localhost:3000/plans/:site where `:site` is the slug of a site with the free plan.
  - Assert that you are taken to checkout with a free trial by clicking 'Start Free Trial'.
  - Assert that you are taken to checkout with a full plan by clicking 'Upgrade Now'.
- Visit http://calypso.localhost:3000/plans/:site where `:site` is the slug of a site with a free trial.
  - Assert that you see the `PlanOverview` component indicating the amount of time remaining for the trial.
  - Assert that you are taken to checkout with the full plan by clicking 'Purchase Now'.
- Visit http://calypso.localhost:3000/plans/:site where `:site` is the slug of a site with a premium/business plan.
  - Assert that the behavior of the buttons below each plan (to manage the plan or upgrade to business if the site has premium) is unchanged.

**Reviews**
- [x] Code
- [x] Product

drewblaisdell commented:
I added the `client/state` files to their own PR: #1660

This PR is now waiting for the result of that review.

scruffian commented:
Forgive me if I have misunderstood, but isn't one of the strengths of Redux that we don't have to pass the state down through the props any more? I guess that's an option we have, but we can still do it the old way too?

scruffian commented on client/components/plans/plan-list/index.jsx:
```diff
@@ -26,28 +28,45 @@ module.exports = React.createClass( {
+			plansList,
+			currentPlan;
+
+		if ( plans.length === 0 || ( ! this.props.isInSignup && ! this.props.sitePlans.hasLoadedFromServer ) ) {
```
This condition looks long and ugly. Is it worth extracting it?

scruffian commented on client/components/plans/plans-compare/index.jsx:
```diff
@@ -168,5 +181,23 @@ module.exports = React.createClass( {
+
+module.exports = connect(
+	( state, props ) => {
+		if ( props.isInSignup ) {
```
Is relying on this prop still the best solution here? Couldn't we do something like looking at the user's current sites, and if they don't have any assume that it's in signup? I guess the problem with that is when users sign up for subsequent sites. I still feel like this prop isn't a great solution though.

drewblaisdell replied:
Good point. I updated this to check for the presence of a selected site.

scruffian commented on client/my-sites/plans/controller.jsx:
```diff
@@ -78,13 +78,15 @@ module.exports = {
-					context={ context } />
-			</CartData>,
+			<ReduxProvider store={ context.store }>
+				<CartData>
```
I assume we still need CartData here?

drewblaisdell replied:
Yep. Cart data isn't stored in the global state tree.

scruffian commented on client/my-sites/plans/controller.jsx:
```diff
@@ -4,6 +4,7 @@
 var page = require( 'page' ),
 	ReactDom = require( 'react-dom' ),
 	React = require( 'react' ),
+	ReduxProvider = require( 'react-redux' ).Provider,
```
You just call this `Provider` below. It probably makes sense to be consistent.

drewblaisdell replied:
I renamed the package `ReduxProvider` in the signup controller.

scruffian commented:
This looks great to me. It would probably be good for @stephanethomas and @Tug to take a look too...

Tug commented:
I'll have a look but don't wait for me to merge this ;)

spncrb commented:
:+1: LGTM everything works as expected

stephanethomas commented on client/components/plans/plan-list/index.jsx:
```diff
@@ -25,29 +27,47 @@ module.exports = React.createClass( {
+				return (
+					<Plan placeholder={ true }
+						isInSignup={ this.props.isInSignup }
+						key={ 'plan-' + n } />
```
Not a huge deal but we could have used a template string here.

stephanethomas commented on client/components/plans/plan-list/index.jsx:
```diff
@@ -25,29 +27,47 @@ module.exports = React.createClass( {
+		if ( ! this.props.isInSignup ) {
+			// check if this site was registered via the JPPHP "Jetpack Start" program
+			// if so, we want to display a message that this plan is managed via the hosting partner
+			currentPlan = find( this.props.sitePlans.data, { currentPlan: true } );
```
I wonder if it would not be the time to introduce a function to retrieve the current plan in the [`plans` module](https://github.com/Automattic/wp-calypso/blob/master/client/lib/plans/index.js). What do you think?

stephanethomas commented:
I'm not sure if this relates to this pull request but I was presented with both paid plans on the `Purchases` page after upgrading from Premium to Business (using the store sandbox):

![screenshot](https://cloud.githubusercontent.com/assets/594356/11917524/d08145d4-a70a-11e5-84d0-cd81d91e8109.png)

stephanethomas commented:
I've just found another issue on the [`Plans` step](http://calypso.localhost:3000/start/plans) during signup where no price would be displayed:

![screenshot](https://cloud.githubusercontent.com/assets/594356/11917584/0478ecf4-a70e-11e5-900b-55b4096f31fb.png)

drewblaisdell commented:
Thanks for catching that, @stephanethomas. I updated and added to the PR to address your feedback.

> I'm not sure if this relates to this pull request but I was presented with both paid plans on the `Purchases` page after upgrading from Premium to Business (using the store sandbox)

This is unrelated to this PR, and is because we don't never remove purchases from `PurchasesStore`. I opened an issue for it: #1888

drewblaisdell commented:
Note: if this is merged after #1780, the changes to the `siteSpecificPlansDetailsList` assembler from that PR will need to be copied into the `sitePlans` assembler in this PR.

Tug commented on client/components/plans/plans-compare/index.jsx:
```diff
@@ -14,19 +15,25 @@ var observe = require( 'lib/mixins/data-observe' ),
 	],
 
+	componentWillReceiveProps: function( nextProps ) {
+		if ( ! this.props.isInSignup && this.props.selectedSite.ID !== nextProps.selectedSite.ID ) {
```
Could use a short comment here to explain in when it happens

Tug commented on client/my-sites/plans/main.jsx:
```diff
@@ -99,3 +113,18 @@ module.exports = React.createClass( {
 } );
+
+module.exports = connect(
+	( state, props ) => {
```
Could we use named functions here? Just to make it easier for us who are starting with redux.

Tug commented:
Code looks good :boom:
